core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -1061,10 +1061,22 @@ public static Expression convert(Expression operand, Type fromType,
           Expressions.convert_(operand, toBox.primitiveClass),
           toBox);
     } else if (fromType == java.sql.Date.class) {
-      if (toBox == Primitive.INT) {
+      if (toType == int.class) {
```

Comment:
There is already similar logic in EnumUtils.java [1], why couldn't we use that ? And there is already method:
```java
 // SqlFunctions.java
  public static Integer toIntOptional(java.util.Date v) {
    return v == null ? null : toInt(v);
  }
```


[1] https://github.com/apache/calcite/blob/620ca17d4e283d58b3ab4c573b911702fd4127c7/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java#L243

Reply:
It is a little different. To some extent, `RexToLixTranslator.convert(Expression operand, Type fromType, Type toType)` is more general.
In `RexToLixTranslator.convert(Expression operand, Type fromType, Type toType)`, the `fromType` is not always the same as `operand.getType()`. Because in current codegen, there are some places (e.g., the row input) need to use `Object.class` as common type. 
For example, the sample code below:
```
Object[] input = ...
//convert(input[0], Timestamp.class, long.class)
long x = SqlFunctions.toLong(input[0]);
```
Therefore, `EnumUtils.enforce` function cannot meet the requirement. This is also the reason why we have `toInt(Object ..)`/`toLong(Object ..)` in `SqlFunctions`.
It is introduced in [CALCITE-1569]((https://issues.apache.org/jira/browse/CALCITE-1569)

Reply:
Can we make some effort to keep one piece of that code, either `EnumUtils.enforce` or `RexToLixTranslator.convert`

Reply:
Yes, EnumUtils.enforce is only used in one place, we can refactor it.

Reply:
There are several places that we need to decide if the data types are among `DATE`, `TIME` and `TIMESTAMP`, can we make some effort to abstract a method in EnumUtils.java like:
```java
representsAsInterType (type) {
 return (type == date || type == time || type == timestamp)
}
```
to decide if we need to handle specifically when we need to do that ?

Reply:
There is a similar function in `RexToLixTranslator`. Can we move all these util functions to `EnumUtils`? I think it can make `RexToLixTranslator` more clean.
```
static boolean isA(Type fromType, Primitive primitive) {
    return Primitive.of(fromType) == primitive
        || Primitive.ofBox(fromType) == primitive;
  }
```

Reply:
I have no strong objections for where to put these tools, either is okey for me.